tools/shared/src/main/scala/org/scalajs/core/tools/io/IRFileCache.scala

```diff
@@ -0,0 +1,314 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2014, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.core.tools.io
+
+import scala.annotation.tailrec
+
+import java.net.URI
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+import org.scalajs.core.ir
+
+/**
+ * Centralized Scala.js IR cache.
+ *
+ * Caches all Scala.js IR used in a given JVM. It supports creating of multiple
+ * sub-caches ([[IRFileCache.Cache]]) that track individual file sets.
+ * The global cache is fully thread safe. However, the sub-caches are not.
+ */
+final class IRFileCache {
+  /* General implementation comment: We always synchronize before doing I/O
+   * (instead of using a calculate and CAS pattern). This is since we assume
+   * that paying the cost for synchronization is lower than I/O.
+   */
+
+  import IRFileCache._
+
+  /** Holds the cached IR */
+  private[this] val globalCache = new ConcurrentHashMap[String, PersistedFiles]
+
+  // Statistics
+  private[this] val statsReused = new AtomicInteger(0)
+  private[this] val statsInvalidated = new AtomicInteger(0)
+  private[this] val statsTreesRead = new AtomicInteger(0)
+
+  /** Create a new sub-cache.
+   *
+   *  Users should call [[Cache.free]] once they are done to allow for more
+   *  aggressive GC.
+   */
+  def newCache: Cache = new Cache
+
+  /** Approximate statistics about the cache usage */
+  def stats: IRFileCache.Stats = {
+    new IRFileCache.Stats(statsReused.get, statsInvalidated.get,
+        statsTreesRead.get)
+  }
```

Comment:
Something we could do here, is expose a `clearStats` method. But then we'd need a way to report and invoke this method in sbt, once "everything is done".
That would actually give even better stats, since they would be over the whole sbt run.